src/main/scala/chiseltest/formal/vips/MagicPacketTracker.scala

```diff
@@ -0,0 +1,109 @@
+package chiseltest.formal.vips
+
+import chisel3._
+import chisel3.experimental.IO
+import chisel3.util._
+
+/** Tracks random packets for formally verifying FIFOs
+  *
+  *  This ensures that when some data enters the FIFO, it will always be dequeued after the correct number of elements.
+  *  So essentially we are verifying data integrity. Note that this does not imply that the FIFO has no bugs
+  *  since e.g., a FIFO that never allows elements to be enqueued would easily pass our assertions.
+  *
+  *  This module was inspired by the MagicPacketTracker used in the evaluation of the following paper:
+  *  Mann, Makai, and Clark Barrett. "Partial order reduction for deep bug finding in synchronous hardware.", CAV'20.
+  */
+object MagicPacketTracker {
+  def apply[D <: Data](enq: ValidIO[D], deq: ValidIO[D], depth: Int, debugPrint: Boolean): Unit = {
+    val tracker = Module(new MagicPacketTracker(chiselTypeOf(enq.bits), depth, debugPrint))
+    tracker.enq := enq
+    tracker.deq := deq
+    val startTracking = IO(Input(Bool()))
+    tracker.startTracking := startTracking
+  }
+  def apply[D <: Data](enq: ValidIO[D], deq: ValidIO[D], depth: Int): Unit =
+    apply(enq, deq, depth, debugPrint = false)
+  def apply[D <: Data](enq: DecoupledIO[D], deq: DecoupledIO[D], depth: Int, debugPrint: Boolean): Unit =
+    apply(asValid(enq), asValid(deq), depth, debugPrint)
+  def apply[D <: Data](enq: DecoupledIO[D], deq: DecoupledIO[D], depth: Int): Unit =
+    apply(asValid(enq), asValid(deq), depth)
+
+  private def asValid[D <: Data](port: DecoupledIO[D]): ValidIO[D] = {
+    val validIO = Wire(ValidIO(chiselTypeOf(port.bits)))
+    validIO.valid := port.fire
+    validIO.bits := port.bits
+    validIO
+  }
+}
+
+class MagicPacketTracker[D <: Data] private (dataTpe: D, fifoDepth: Int, debugPrint: Boolean) extends Module {
+  require(fifoDepth > 0, "Fifo depth needs to be positive!")
+  val enq = IO(Input(ValidIO(dataTpe)))
+  val deq = IO(Input(ValidIO(dataTpe)))
+
+  // count the number of elements in the fifo
+  val elementCount = RegInit(0.U(log2Ceil(fifoDepth + 1).W))
+  val nextElementCount = Mux(
+    enq.fire && !deq.fire,
+    elementCount + 1.U,
+    Mux(!enq.fire && deq.fire, elementCount - 1.U, elementCount)
+  )
+  elementCount := nextElementCount
+
+  // track a random "magic" packet through the fifo
+  val startTracking = IO(Input(Bool()))
+  val isActive = RegInit(false.B)
+  val packetValue = Reg(chiselTypeOf(enq.bits))
+  val packetCount = Reg(chiselTypeOf(elementCount))
+
+  when(!isActive && enq.fire && startTracking) {
+    when(deq.fire && elementCount === 0.U) {
+      assert(
+        enq.bits.asUInt === deq.bits.asUInt,
+        "element should pass through the fifo, but %x != %x",
+        enq.bits.asUInt,
+        deq.bits.asUInt
+      )
+    }.otherwise {
+      isActive := true.B
+      packetValue := enq.bits
+      packetCount := nextElementCount
```

Comment:
nextElementCount in this context is enq.fire && !deq.fire so is it OK to store elementCount directly instead? (and change the check below to perform the assertion on packetCount === 0.U)?

Reply:
We could. Not sure if that would make it easier to understand though.